### ISSUE TITLE
Confusion about steps, samples, and trajectories.

### ISSUE THREAD

**Issue by IdsTeepe** [openpathsampling/openpathsampling#868]
As far as I understand, every step has a sample set with (at least?) one sample, and every sample has a trajectory associated with it. Taking a simple `OneWayShootingMoveScheme` on an alanine dipeptide, then running `sampler.run(500)` I get the following results:

```
print(len(sampler.storage.steps), 
      len(sampler.storage.samples),
      len(sampler.storage.trajectories))
-------------------------------------------------
501 501 501
```

Taking a look at the steps, we see that they have the last accepted sample associated with them:
```
for step in sampler.storage.steps:
    print(step.idx(sampler.storage),
          step.active.samples[0].idx(sampler.storage),
          len(step.active.samples),
          step.change.accepted)
all([(not step.change.accepted or \
     (step.idx(sampler.storage) == step.active.samples[0].idx(sampler.storage)))
        for step in sampler.storage.steps])
-------------------------------------------------
0 0 1 True
1 1 1 True
2 1 1 False
3 1 1 False
4 4 1 True
5 4 1 False
     ...
496 496 1 True
497 497 1 True
498 498 1 True
499 498 1 False
500 498 1 False
True
```

Similarly, all samples have a unique trajectory associated with them:
```
for sample in sampler.storage.samples[:6]:
    print(sample.idx(sampler.storage),
          sample.trajectory.idx(sampler.storage))
all([(sample.idx(sampler.storage) == sample.trajectory.idx(sampler.storage))
    for sample in sampler.storage.samples])
-------------------------------------------------
0 0
1 1
2 2
3 3
4 4
5 5
True
```
(The documentation confirms this last one:)
> Since each Sample is a single representative of a single ensemble, each Sample consists of one replica ID, one trajectory, and one ensemble.

The first part of my question (or personal issue) is thus: Is this the correct way to think about steps, samples and trajectories in `openpathsampling`?

Recently, I've been doing some runs not with `OneWayShootingMoveScheme` but with [`spring_shooting`](https://e-cam.readthedocs.io/en/latest/Classical-MD-Modules/modules/OpenPathSampling/ops_spring_shooting/readme.html); and this is where my confusion started:
In this move scheme, a snapshot is chosen within a frame `delta_max` of the previous chosen index:
```
        index = previous_shooting_index + (dframe - delta_max)

        # Check if a valid index has been selected
        if index < 0 or index >= len(trajectory):
            self.acceptable_snapshot = False
            index = 1  # Needed to prevent selecting an out of range index
        else:
            self.acceptable_snapshot = True
```

`openpathsampling` runs fine combining the Gromacs engine with `spring_shooting`, and every step still has one sample (the previous accepted sample). However, not every sample has a trajectory:
```
print(len(sampler.storage.steps), 
      len(sampler.storage.samples),
      len(sampler.storage.trajectories))
all([(sample.idx(sampler.storage) == sample.trajectory.idx(sampler.storage))
    for sample in sampler.storage.samples])
-------------------------------------------------
501 501 473
False
```
(The above statement is misleading: Not every sample has a _unique_ trajectory. In this case, both samples 218 and 219 have trajectory 218 associated with them, for example.)

This causes me some trouble when during analysis: Let's say I want to plot the CVs of the last accepted step (№ 493), then I'll have to use `sampler.storage.trajectories[465]`
(Previously, I'd manually count the amount of samples with no trajectory preceding [or succeeding] the wanted sample. During writing of this issue, I notice I can just use `Sample.trajectory.idx()` to obtain the correct value.)
Despite the epiphany I just had, it seems unnatural to me to have sample 493 associated with trajectory 465.
In case the step gets rejected before Gromacs even starts (Should this ever be the case, even?), shouldn't a `None` be added to `storage.trajectories`? This way, everything would stay in line.

Sorry for the long post, writing it cleared out a whole lot of my initial confusion. this misalignment between samples and trajectories is less annoying to me now that I've found out I can use `Sample.trajectory.idx()`, yet I'll still bring up the issue as it seems strange and unnatural to me.

**Comment by dwhswenson** [comment]
No worries about the long post, but a long post gets a long response! ;-)

First off, for the most part, you should **never** go directly to `storage.trajectories` or `storage.samples` for analysis. Among other things, you have no way of know which are accepted and which are rejected if you do that! And in principle, you might have trajectories from multiple simulations (e.g., TPS and committor) in the same file. That wouldn't be recommended, but it is possible, and you would have no way to distinguish between the simulations.

Instead, do your analysis by iterating over the `steps`. By doing that, you're basically replaying the simulation, and you can tell which trajectories were accepted or not (and perhaps do different analysis based on each case!) The "current state" of the system can be found in the `SampleSet` at `step.active` -- these repeat when a step is rejected, giving the correct statistical sampling (and since the repeats are just pointers, we don't worry about the disk space). You can select a `Sample` from the `SampleSet` by ensemble or by replica ID -- for TPS, the replica ID is usually 0 (unless you told it to be something else), so that's usually the easiest way: `step.active[0].trajectory`. You can get the trial `Sample`s with `step.change.trials`, and you can see whether or not they were accepted with `step.change.accepted`. The `MCStep` object has an attribute called `mccycle` which gives you the step number, and it can also figure out the `PathSimulator` that generated it (so even if you put two simulations in one file, you can tell what came from which.) 

The figure at http://openpathsampling.org/latest/topics/data_objects.html (there's a similar one in the first OPS paper) is designed to give an overview of this. That page also gives the best description (so far) of how the hierarchy of data objects objects works.

A few specific comments:

> every step has a sample set with (at least?) one sample

Stating carefully: an `MCStep` contains `active`, which is a `SampleSet` and which should contain as many `Sample`s as are "live" in your system (for TIS, that might be one per interface). In everything we've implemented, `len(active)` is constant over the steps for a given simulation, but that's not required.

However, I think you're actually asking about how many *new* samples there are; i.e., how many samples are in `step.change.trials` (which is actually a `list` of `Sample`s, not a `SampleSet` -- I understand why that would be confusing, but there are (long-winded) reasons!)

I think there actually *can* be 0 trial samples, although I'm not sure (I think that the dummy moves that were used for single-replica TIS may have created no samples, but that's a pretty unique case). A step can definitely have more than one trial sample: a replica exchange move must generate one `Sample` for each replica, so it has 2.

> Not every sample has a unique trajectory. In this case, both samples 218 and 219 have trajectory 218 associated with them, for example.

Careful here: each sample contains exactly one trajectory. But that mapping is not one-to-one -- as a trivial example, replica exchange takes the existing trajectories, and tries them in different ensembles, making new `Sample`s.

For the specifics of spring shooting, though, I am a bit surprised that you get two different samples with the same trajectory in the same ensemble. I'm guessing this has to do with how Sander implemented early rejection (below), but I don't know.

> Let's say I want to plot the CVs of the last accepted step (№ 493)

Assuming you only have results of one simulation in `storage` (or that you're interested in the most recent one), then the last accepted trajectory for replica 0 is `storage.steps[-1].active[0].trajectory`, even if the last step was rejected. `GromacsEngine`-specific: if you want to figure out the TRR files for the trajectory, I think you can get it from `snapshot.filename` (`file_name`?) for the snapshots. For 1-way shooting approaches, one trajectory may include parts of several TRR files!

Fun thought: Using the Gromacs engine, it's really easy to calculate "number of files contributing to each step's trajectory", something like:

```python
active_files_per_step = [len(set(snap.filename for snap in step.active[0].trajectory))
                         for step in steps]
```

I'd be really curious how those distributions differ for one-way shooting and spring shooting (I'm assuming you're comparing those methods). This is another way of looking at the same idea as the number of MC steps between decorrelated trajectories.

> In case the step gets rejected before Gromacs even starts (Should this ever be the case, even?)

Look at that section of the spring shooting code that you identified: if you try to select a shooting index that is less than 0 or greater than the length of the trajectory, you *can't* use that as a shooting point, right? (There's no snapshot to shoot from.) Rejecting the move seems reasonable, and with some math (presumably in the spring shooting paper) you can show that it is also correct, i.e., preserves detailed balance. So you reject the move without making a trajectory.

---

tl;dr: The confusion is over the concept that any of the stores other than `steps` are intended to be used for analysis. They're really just places to store things, and we try to keep them efficient (so we don't pad with extra stuff). When doing analysis, loop over `steps`: from there, you can get anything, and you have access to all the context (like acceptance, etc.)

If there's still anything in the docs where we loop over `storage.trajectories` or `storage.sample_sets`, please let me know. In very early versions of the code (early 2015), we did that, but we should get rid of anything that still suggests that's a good idea. I know of one example (committor setup) where I draw straight from `snapshots`, but even there I say it's a bad/lazy way to do things (and I'm still too lazy to redo it properly).

**Comment by IdsTeepe** [comment]
This post certainly cleared up a lot for me! Using `storage.steps[-1].active[0].trajectory` works fine for me, thank you.

One thing about this is still bothering me:
> Look at that section of the spring shooting code that you identified: if you try to select a shooting index that is less than 0 or greater than the length of the trajectory, you can't use that as a shooting point, right? (There's no snapshot to shoot from.) Rejecting the move seems reasonable, and with some math (presumably in the spring shooting paper) you can show that it is also correct, i.e., preserves detailed balance. So you reject the move without making a trajectory.

Right now I'm sampling a system which sometimes has a trajectory of ~150 frames, and at other steps has a trajectory of ~2000 frames. Either I need to pick a very low Δmax, or the spring shooting is guaranteed to often pick an out-of-bounds frame when the trajectory is a lot smaller. For the simulation this doesn't matter a lot, as just picking a new point is not a difficult task. For analysis however, taking all these as _rejected moves_ really skews the acceptance ratio. 
For my most recent run, I would have an acceptance ratio of 15/104 = 14.4%.
Taking into account that 43 of the 89 rejections were due to the spring shooting picking a frame that's outside of the trajectory, however, it seems more reasonable to take the acceptance as 15/(104-43) = 24.5%. To get to this number, however, I need to identify why the move was rejected for all rejected moves, and count them.
Isn't it better to just pick a new start without generating a new step?
In the end it doesn't really matter. It's just an analysis-issue and the sampling works fine.

**Comment by dwhswenson** [comment]
Keep in mind that you're sampling trajectories from a statistical ensemble. Each trajectory has a certain weight associated with it. In order to correctly estimate that weight, the rejections need to be counted as they are.

I don't know the details of how Sander implemented spring shooting, but there should be some way to figure out when no new trial trajectory was run, based on information in the `change`. (We save *everything*, so you can always figure this out.)

It sounds like the bigger question you're asking is "how do I compare the *efficiency* of two algorithms." In general, we usually assume that (statistically valid) algorithms that do more exploration of the path space are more efficient. (Depending on what you're trying to do, this may or may not be correct, but it's usually a good start.)

Measuring how far you've explored in a high-dimensional space is hard. It's reasonably hard to measure even for individual MD frames, and it's *really* hard to measure for trajectories.

You're probably thinking of using acceptance probability as a proxy for exploration, based on the reasonable argument that if more trials are accepted, then you've probably explored more of the trajectory space. (Note that there's a limit to that argument anyway: if the trial generates a very small change, then you have a high acceptance ratio, but didn't explore very far.)

In the spring shooting paper, they got around the problem you've noticed by introducing the least-changed path as a metric for efficiency. I think that Sander has implemented the LCP for OPS, but I'm not sure where that code is.

**Comment by sroet** [comment]
>  I think that Sander has implemented the LCP for OPS, but I'm not sure where that code is.

Some (very) preliminary code lives [here](https://gitlab.e-cam2020.eu/sroet/least_changed_path)
